python/social_bot/parser.py

```diff
@@ -0,0 +1,125 @@
+
+    A word class can be added using ``add_word_class()``. And a parse rule can
+    be added using ``add_rule()``. A parse rule is a python regular expression
+    with a special syntax to denote slot. A slot can be denoted using the
+    following two formats:
+
+    * <WORD_CLASS_NAME>: the value of this slot is from the given word class.
+      The name of the slot is also WORD_CLASS_NAME.
+    * <WORD_CLASS_NAME:SLOT_NAME>: the value of this slot is from the given word
+      class. The name of the slot is SLOT_NAME.
+
+    All the slot names of a single parse rule need to be different.
+
+    Examples:
+
+    .. code-block:: python
+
+        parser = ReParser()
+        parser.add_word_class("OBJ", ["apple", "orange", "peach"])
+        parser.add_rule("an? <OBJ:OBJ_RIGHT> is on the right of an? <OBJ:OBJ_LEFT>", "OBJ_POS")
+        parser.add_rule("an? <OBJ:OBJ_LEFT> is on the left of an? <OBJ:OBJ_RIGHT>", "OBJ_POS")
+
+        result = parser.parse("a peach is on the right of an apple")
+        assert result.rule_name == "OBJ_POS"
+        assert result.slot_values == {"OBJ_RIGHT": "peach", "OBJ_LEFT": "apple"}
+
+    """
+
+    def __init__(self):
+        self._word_classes = {}
+        self._rules = []
+
+    def add_rule(self, template, rule_name=''):
+        """Add a parse rule.
+
+        Args:
+            template (str): template for the sentences to be parsed by this rule.
+            rule_name (str): name of the parse rule.
+        """
+        word_classes = []
+        slot_names = []
+        new_template = ""
+        while True:
+            m = re.search(r"<\w+(:\w+)?>", template)
+            if not m:
+                break
+            slot = m.group(0)
+            slot = slot[1:-1]
+            if ':' in slot:
+                word_class, slot_name = slot.split(':')
+            else:
+                word_class, slot_name = slot, slot
+            assert word_class in self._word_classes, "Unknown world class %s" % word_class
+            word_classes.append(word_class)
+            assert slot_name not in slot_names, "duplicated slot name %s" % slot_name
+            slot_names.append(slot_name)
+            new_template += template[:m.start()] + ("(?P<%s>(%s))" % (
+                slot_name, '|'.join(self._word_classes[word_class])))
+            template = template[m.end():]
+        pattern = re.compile(new_template)
+        self._rules.append(
+            ParseRule(pattern, rule_name, word_classes, slot_names))
+
+    def add_word_class(self, word_class_name, words):
+        """Add a word class.
+
+        Args:
+            word_class_name (str): name of the word class
+            words (list[str]): the words/phrases of this word class.
+        """
+        assert word_class_name not in self._word_classes, (
+            "duplicated world class name %s" % word_class_name)
+        self._word_classes[word_class_name] = words
+
+    def parse(self, sentence):
```

Comment:
probably should call .lower() on sentence?

Reply:
Since all the words are from word IDs given the known vocabulary, they should not have wrong case.